File: 2019/go/day1.go
package main

import (
    "fmt"
    "os"
    "strings"
    "strconv"
)

func day1() {
    input, err := os.ReadFile("../input/day1.txt")
    must(err)
    p1, p2 := d1(string(input))
    fmt.Println(p1)
    fmt.Println(p2)
}

func d1(input string) (int, int) {
    words := strings.Fields(input)
    var sum int
    var extra int

    for _, word := range words {
        i, err := strconv.Atoi(word)
        must(err)

        fuel := (i / 3) - 2
        sum += fuel
        extra += extraFuelMass(fuel)
    }

    return sum, extra
}

func extraFuelMass(fuel int) int {
    extra := fuel
    sum := 0

    for extra > 0 {
        sum += extra 
        extra = (extra / 3) - 2
    }

    return sum
}

File: 2019/cpp/src/day1.cpp
#include "days.hpp"
#include "util.hpp"

#include <iostream>

int extra_fuel(int fuel) {
    int sum = 0;
    int extra = fuel;

    while (extra > 0) {
        sum += extra;
        extra = (extra / 3) - 2;
    }

    return sum;
}

void solve_d1(std::vector<std::string>& input, int& p1, int& p2) {
    int sum = 0;
    int extra = 0;

    for (std::string word : input) {
        int weight = stoi(word);
        int fuel = (weight / 3) - 2;
        sum += fuel;
        extra += extra_fuel(fuel);
    }

    p1 = sum;
    p2 = extra;
}

void day1() {
    std::vector<std::string> input = read_input_lines(1);
    int p1, p2;

    solve_d1(input, p1, p2);

    std::cout << p1 << ", " << p2 << std::endl;
}
